Support sub-menu for dropdown

jonahtanjz opened the pull request:
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [x] Feature addition or enhancement
- [ ] Code maintenance
- [ ] Others, please explain:

<!--
  If this pull request is addressing an issue, link to the issue: "Fixes #xxx" or "Resolves #xxx"

  If the pull request completely addresses the issue, use one of the issue closing keywords. https://docs.github.com/en/enterprise/2.16/user/github/managing-your-work-on-github/closing-issues-using-keywords

  Otherwise, elaborate further on the rationale of this pull request as needed
-->

**Overview of changes:**
Resolves #212.

Sub-menus can be added by nesting dropdowns so that users do not have to remember a new syntax. A new component `<submenu>` is created to handle the logic. `<dropdown>` component is edited such that if a `<dropdown>` is nested, it will be converted to a `<submenu>`. Sub-menus can also contain sub-menus using the same syntax for multiple levels. Currently, sub-menus will appear to the right of a dropdown on desktop version and becomes an accordion on mobile.

**Anything you'd like to highlight / discuss:**
Implementation.

**Testing instructions:**
Sub-menus can be added by nesting `<dropdown>`.

For example:
```
<dropdown>
    <li>...</li> // normal dropdown item
    <li>...</li>
    <dropdown> // will be converted to submenu
        <li>...</li> // submenu item
        <li>...</li>
        <dropdown> // 2nd-level submenu
            <li>...</li> // 2nd-level submenu item
            <li>...</li>
        </dropdown>
    </dropdown>
</dropdown>

```
A dropdown with 2 levels of submenu will be created.

<!--
  Any special testing instructions in not including our automated tests.
-->

**Proposed commit message: (wrap lines at 72 characters)**
Add sub-menu support for dropdown
<!--
  See this link for more info on how to write a good commit message:
  https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message

  As best as possible, write a succinct commit title in 50 characters
-->

---

**Checklist:** :ballot_box_with_check:

<!--
  Prefix your pull request title with [WIP] if any of these are not complete yet

  Tick non-applicable items as well
-->

- [x] Updated the documentation for feature additions and enhancements
- [x] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No blatantly unrelated changes
- [x] Pinged someone for a review!

jonahtanjz commented:
@ang-zeyu this is ready for review and do let me know on how this can be improved on. If the implementation is fine, will proceed to add in tests and documentation as well.

ang-zeyu reviewed (commented):
Just a review on the overall approach / issues:

1. I was thinking of using the https://v3.vuejs.org/guide/component-provide-inject.html#working-with-reactivity. 

e.g.

you could do
```
<dropdown>
  <div class="some-wrapper">
    <dropdown></dropdown> // would still work when nested
  </div>
</dropdown>
```

the approach here is more consistent with vue-strap's, but as you might have observed its a rather low level approach that ties itself heavily to the dom. (note vue-strap didn't have this api at the time)

But I would be ok with accepting this as quite a bit of work has been done already. If you're up for it, let's change this over to provide/inject (could even make a separate PR to shift the original implementation to it first if it makes things easier)

2. Positioning of submenus needs to be dynamic (we can assume right position first then flip it over to the left otherwise). (note the submenu may overflow over the screen currently) You could explore the [popperjs](https://popper.js.org/) library to do this (bootstrap-vue uses this internally for its popovers,dropdowns...)

3. Really small nit: Make the submenu show on-hover as well? 🤔

ang-zeyu commented:
One more note since you're tackling #980 as well: you may want to finish shifting the mobile navbar design to it first, as dropdowns contained within navbars look / behave quite differently - less work to do here

jonahtanjz commented:
@ang-zeyu I've made some changes accordingly.
> I was thinking of using the https://v3.vuejs.org/guide/component-provide-inject.html#working-with-reactivity.

I didn't use provide/inject in the end as when I was trying to implement this, I managed to remove all the unnecessary communication between the `dropdown` and `submenu` components. Not sure where else to use it as the rest of the search of the virtual DOM seems necessary, for example listening for clicks on `li>a` as these are user defined.

> Positioning of submenus needs to be dynamic (we can assume right position first then flip it over to the left otherwise). (note the submenu may overflow over the screen currently) You could explore the popperjs library to do this (bootstrap-vue uses this internally for its popovers,dropdowns...)

I tried out popperjs but the auto positioning didn't work very well/a little buggy. I switched to using a modified bootstrap javascript which similarly solves this issue. Hopefully that's fine ;)

I've also added tests and updated the docs.

ang-zeyu commented on packages/vue-components/src/Submenu.vue:
```diff
@@ -0,0 +1,177 @@
+      type: Boolean,
+      default: false,
+    },
+    'class': null,
```
👀

jonahtanjz replied:
Thanks! Updated :)

ang-zeyu commented on packages/vue-components/src/Submenu.vue:
```diff
@@ -0,0 +1,177 @@
+  },
+  computed: {
+    classes() {
+      return [
```
let's place this directly in the template only for consistency with other components'

jonahtanjz replied:
Alright. Updated :)

ang-zeyu commented on docs/userGuide/syntax/dropdowns.mbdf:
```diff
@@ -56,6 +56,23 @@
+<include src="codeAndOutput.md" boilerplate >
+<variable name="highlightStyle">html</variable>
+<variable name="code">
+<!-- Can simply nest the dropdown syntax -->
```
```suggestion
<!-- Nest the dropdown syntax to create dropdown submenus -->
```

jonahtanjz replied:
Updated :)

ang-zeyu commented on packages/vue-components/src/Navbar.vue:
```diff
@@ -126,9 +126,19 @@ export default {
+    },
+    addClassIfSubmenu(a, li) {
+      let el = a.parentElement;
+      while (el != li) {
```
`!==`

jonahtanjz replied:
Thanks! Updated :)

ang-zeyu commented on packages/vue-components/src/Submenu.vue:
```diff
@@ -0,0 +1,177 @@
+  },
+  methods: {
+    hideSubmenu() {
+      this.show = false;
```
let's use `data()` for properties that could change

jonahtanjz replied:
Updated :)

ang-zeyu commented on packages/vue-components/src/Submenu.vue:
```diff
@@ -0,0 +1,177 @@
+}
+
+@media (max-width: 767px) {
+  .dropdown-submenu > ul {
```
hmm can this be moved to the same media query below?

jonahtanjz replied:
Thanks! Updated :)

ang-zeyu reviewed (commented):
> I didn't use provide/inject in the end as when I was trying to implement this, I managed to remove all the unnecessary communication between the `dropdown` and `submenu` components.

Not so much for removing unnecessary communication but just allowing parent / child communication across different [compilation scopes](https://vuejs.org/v2/guide/components-slots.html#Compilation-Scope). 

Nice work isolating the implementation as much as possible - only have one not-so-beneficial suggestion where this can be used now. (see comment below)

You could take a look at the following for examples how it could be used more beneficially:
- [bv components](https://github.com/bootstrap-vue/bootstrap-vue/blob/dev/src/components/dropdown/dropdown-item.js)
- our own `<quiz>` / `<question>` / `<q-option>` components

> Not sure where else to use it as the rest of the search of the virtual DOM seems necessary, for example listening for clicks on `li>a` as these are user defined.

yes this is still necessary unfortunately, due to our use of standard dom elements.

> I tried out popperjs but the auto positioning didn't work very well/a little buggy. I switched to using a modified bootstrap javascript which similarly solves this issue. Hopefully that's fine ;)

Seems ok too 👍 

---


Some strange behaviours:

![dropdown-sub](https://user-images.githubusercontent.com/3306138/105692994-20ebd380-5f3a-11eb-926c-062c38f0b51f.gif)
 (main / submenu closing on hover)

![Untitled](https://user-images.githubusercontent.com/3306138/105693256-79bb6c00-5f3a-11eb-850c-740cf92f3568.png)
(this is on mobile view)

Some suggestions:

jonahtanjz commented:
@ang-zeyu Thanks for the suggestions and clarification 😄. I've updated it accordingly.

> Some strange behaviours:
> (main / submenu closing on hover)

For the main menu, it is closing because when the inner children of the submenu toggle is clicked/hovered, it triggers an `blur` event. I've fixed this by preventing the event from bubbling up. 

As for the submenu closing on hover, it is because when the user hovers over the submenu toggle, it will mirror a click. That is the reason why the second hover will close the submenu. One workaround is to only allow submenus to be opened by hovering over and not by clicking. This is the current implementation for now.

jonahtanjz commented on packages/vue-components/src/Dropdown.vue:
```diff
@@ -72,6 +79,7 @@ export default {
       return toBoolean(this.disabled);
     },
     isLi () { return this.$parent._navbar || this.$parent.menu || this.$parent._tabset },
+    isSubmenu() { return this.$parent._dropdown || this.$parent._submenu },
```
Thanks for the clarification. Have added this in :)

jonahtanjz commented on packages/vue-components/src/Dropdown.vue:
```diff
@@ -13,6 +13,13 @@
     </slot>
   </li>
+  <submenu v-else-if="isSubmenu" ref="submenu">
+    <slot slot="button" name=button></slot>
```
Thanks! Updated :)

ang-zeyu commented on packages/vue-components/src/Dropdown.vue:
```diff
@@ -112,7 +120,7 @@ export default {
       this.show = true;
       $(this.$refs.dropdown).findChildren('ul').each(ul => ul.classList.toggle('show', true));
-    },
+    }
```
let's correct these in another PR; we could enable linting for `dropdown.vue` as well. (the unlinted ones are files that haven't changed much from vue-strap)

jonahtanjz replied:
Alright sure will open another PR for this 👍

ang-zeyu commented on packages/vue-components/src/Submenu.vue:
```diff
@@ -0,0 +1,183 @@
+    disabledBool() {
+      return toBoolean(this.disabled);
+    },
+    showBool() {
```
this (could replace with just `show`) and `slots()` dosen't seem used

jonahtanjz replied:
Thanks!. Replaced all `showBool` with `show`

ang-zeyu commented on packages/vue-components/src/Submenu.vue:
```diff
@@ -0,0 +1,183 @@
+</script>
+
+<style scoped>
+.dropdown-submenu {
```
let's indent these to fit the stylelint rules too (they've never been validated in our scripts for `.vue` files but should be)

jonahtanjz replied:
Updated :)

ang-zeyu reviewed (commented):
Couple more nits:

jonahtanjz commented:
@ang-zeyu PR is updated accordingly :)

ang-zeyu reviewed (approved):
Lgtm 👍
